core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -118,7 +147,13 @@ class HadoopMapReduceCommitProtocol(
     }
 
     dir.map { d =>
-      new Path(new Path(stagingDir, d), filename).toString
+      if (dynamicPartitionOverwrite && isSpeculationEnabled) {
```

Comment:
I think this issue will happen even when speculation is not enabled. When the task fails and writes partial data and when the new attempt executes, it fails with FileAlreadyExistException